### PR TITLE
Concatenate multiple cookies with semicolon

### LINKED ISSUE #42856: Cookie header values should be separated by semi-colon not comma.
[RFC HTTP State Management Mechanism](https://tools.ietf.org/html/rfc6265#section-4.2.1)
Grammar:
```
   cookie-header = "Cookie:" OWS cookie-string OWS
   cookie-string = cookie-pair *( ";" SP cookie-pair )
```

We already properly concatenated within `SocketsHttpHandler` when we merge request headers with cookies from the container: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L247

We should do the same for all `Cookie` header values.

Repro:
```C#
var client = new HttpClient();
var message = new HttpRequestMessage(HttpMethod.Get, "http://localhost:5001");

// All of the Cookie header bellow will get concatenated with ','.
message.Headers.Add("Cookie", new [] {"test1=1","test2=2"});
message.Headers.Add("Cookie", "test3=3");
message.Headers.TryAddWithoutValidation("Cookie", new [] {"test4=4", "test5=5"});
message.Headers.TryAddWithoutValidation("Cookie", "test6=6");
var response = await client.SendAsync(message);
```
Wireshark snip:
image

Upon first look, adding a custom `HttpHeaderParser` overriding the separator and using it for `Cookie` in known header definitions: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs#L43 should do the trick.

Original issue: https://github.com/microsoft/reverse-proxy/issues/437

### PR DESCRIPTION
**Pull request by devsko**
Fixes #42856 

According to [RFC 6265](https://datatracker.ietf.org/doc/html/rfc6265#section-4.2) multiple cookies must be concatenated with `";" SP` when writing the `Cookie` header.
The existing implementation complies to this when writing the cookies that are added to a `CookieContainer` but all cookies that are added directly to the `HttpRequestHeaders` collection are concatenated with a comma. 
This PR ensures that all cookies are concatenated with a semicolon.

### CODE REVIEW

**Comment by devsko** [comment]
Okay. The build fails because the adapted tests are also running against `WinHttpHandler` on net48.
I assume it is not an option to build `System.Net.Http` targeting .NET Framework with changed behavior, right?

Then I would suggest to ifdef the tests that run also on net48
```csharp
#if NETFRAMEWORK
     var separator = ", ";
#else
     var separator = "; ";
#endif
```
@ManickaP what do you think?

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs [hunk 1]
```diff
@@ -40,7 +40,7 @@ internal static class KnownHeaders
         public static readonly KnownHeader ContentRange = new KnownHeader("Content-Range", HttpHeaderType.Content | HttpHeaderType.NonTrailing, GenericHeaderParser.ContentRangeParser, null, H2StaticTable.ContentRange);
         public static readonly KnownHeader ContentSecurityPolicy = new KnownHeader("Content-Security-Policy", http3StaticTableIndex: H3StaticTable.ContentSecurityPolicyAllNone);
         public static readonly KnownHeader ContentType = new KnownHeader("Content-Type", HttpHeaderType.Content | HttpHeaderType.NonTrailing, MediaTypeHeaderParser.SingleValueParser, null, H2StaticTable.ContentType, H3StaticTable.ContentTypeApplicationDnsMessage);
-        public static readonly KnownHeader Cookie = new KnownHeader("Cookie", H2StaticTable.Cookie, H3StaticTable.Cookie);
+        public static readonly KnownHeader Cookie = new KnownHeader("Cookie", HttpHeaderType.Request, CookieHeaderParser.Parser, null, H2StaticTable.Cookie, H3StaticTable.Cookie);
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs:43]: Note to reviewers: Adding `HttpHeaderType.Request` is a breaking change as you can no longer add this header to the content/response collections with validating APIs.

We may still want to do it, just making sure it's a deliberate decision.

Comment (devsko) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: @MihaZupan I changed the header type to `HttpHeaderType.Custom` for now. Thanks

**Comment by ManickaP** [comment]
> Okay. The build fails because the adapted tests are also running against `WinHttpHandler` on net48. I assume it is not an option to build `System.Net.Http` targeting .NET Framework with changed behavior, right?
> 
> Then I would suggest to ifdef the tests that run also on net48
> 
> ```cs
> #if NETFRAMEWORK
>      var separator = ", ";
> #else
>      var separator = "; ";
> #endif
> ```
> 
> @ManickaP Marie Píchová FTE what do you think?

Yeah, we're not gonna service framework for this, so #if #else is fine solution for this.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Cookies.cs [hunk 2]
```diff
@@ -262,6 +266,7 @@ public async Task GetAsync_SetCookieContainerAndMultipleCookieHeaders_BothCookie
                     HttpRequestData requestData = await serverTask;
                     string cookieHeaderValue = GetCookieValue(requestData);
 
+#if NETFRAMEWORK
                     // Multiple Cookie header values are treated as any other header values and are
```
Comment (ManickaP) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Cookies.cs:270]: I'm leaning towards simplifying this to:
```C#
#if NETFRAMEWORK
var separators = new string[] { "; ", ", " };
#else
var separators = new string[] { "; " };
#endif
var cookieValues = cookieHeaderValue.Split(separators, StringSplitOptions.None);
...
```
We can keep the comment about cookie container and how the mixed separators got in.
What do you think?

**Review by ManickaP — APPROVED** [review]
Looks good, one optional suggestion. Thank you for fixing this!

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Cookies.cs (hunk 2 above)
Comment (devsko) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Cookies.cs]: Sure, fixed.

**Comment by devsko** [comment]
Build failures seem unrelated.

**Pull request merged**